File: Classes/Controllers/PBGitWindowController+Dialogs.m
#import "PBGitWindowController.h"

#import "PBCommitHookFailedSheet.h"
#import "PBError.h"
#import "PBGitCommitController.h"
#import "PBGitDefaults.h"
#import "PBGitRepository.h"
#import "PBGitXMessageSheet.h"
#import "GitX-Swift.h"

#pragma clang diagnostic push
// These methods remain declared on the stable primary interface.
#pragma clang diagnostic ignored "-Wobjc-protocol-method-implementation"
@implementation PBGitWindowController (PBDialog)

- (IBAction)showRepositorySettings:(id)sender
{
	[PBRepositorySettingsController beginSheetForRepository:self.repository windowController:self];
}

- (void)showCommitHookFailedSheet:(NSString *)messageText infoText:(NSString *)infoText commitController:(PBGitCommitController *)controller
{
	[PBCommitHookFailedSheet beginWithMessageText:messageText
										 infoText:infoText
								 commitController:controller
								completionHandler:^(__unused id sheet, NSModalResponse response) {
									if (response == NSModalResponseOK) [self.commitViewController forceCommit:self];
								}];
}

- (void)showMessageSheet:(NSString *)messageText infoText:(NSString *)infoText
{
	[PBGitXMessageSheet beginSheetWithMessage:messageText info:infoText windowController:self];
}

- (void)showErrorSheet:(NSError *)error
{
	if ([error.domain isEqualToString:PBGitXErrorDomain]) {
		[PBGitXMessageSheet beginSheetWithError:error windowController:self];
	} else {
		[[NSAlert alertWithError:error] beginSheetModalForWindow:self.window
											   completionHandler:^(__unused NSModalResponse response){
											   }];
	}
}

- (BOOL)confirmDialog:(NSAlert *)alert suppressionIdentifier:(NSString *)identifier forAction:(void (^)(void))actionBlock
{
	NSParameterAssert(alert);
	__block BOOL didAct = YES;
	if (identifier && [PBGitDefaults isDialogWarningSuppressedForDialog:identifier]) {
		actionBlock();
		return didAct;
	}
	alert.showsSuppressionButton = YES;
	[alert beginSheetModalForWindow:self.window
				  completionHandler:^(NSModalResponse response) {
					  if (response != NSAlertFirstButtonReturn) {
						  didAct = NO;
						  return;
					  }
					  if (identifier && alert.suppressionButton.state == NSControlStateValueOn) [PBGitDefaults suppressDialogWarningForDialog:identifier];
					  actionBlock();
				  }];
	return didAct;
}

@end
#pragma clang diagnostic pop

File: Classes/Controllers/PBGitWindowController+Dialogs.swift
import AppKit

enum WindowDialogPresenter {
    static func showRepositorySettings(for windowController: PBGitWindowController) {
        guard let repository = windowController.repository else { return }
        RepositorySettingsController.beginSheet(for: repository, windowController: windowController)
    }

    static func showCommitHookFailedSheet(
        _ messageText: String,
        infoText: String,
        retryHandler: (() -> Void)?,
        for windowController: PBGitWindowController
    ) {
        PBCommitHookFailedSheet.begin(
            withMessageText: messageText,
            infoText: infoText,
            windowController: windowController
        ) { _, response in
            if response == .OK {
                retryHandler?()
            }
        }
    }

    static func showMessageSheet(
        _ messageText: String,
        infoText: String,
        for windowController: PBGitWindowController
    ) {
        PBGitXMessageSheet.begin(
            withMessage: messageText,
            info: infoText,
            windowController: windowController
        )
    }

    static func showErrorSheet(_ error: Error, for windowController: PBGitWindowController) {
        let nsError = error as NSError
        if nsError.domain == PBGitXErrorDomain {
            PBShowGitXErrorSheet(nsError, windowController)
        } else if let window = windowController.window {
            NSAlert(error: nsError).beginSheetModal(for: window) { _ in }
        }
    }

    static func confirmDialog(
        _ alert: NSAlert,
        suppressionIdentifier identifier: String?,
        for windowController: PBGitWindowController,
        action actionBlock: @escaping () -> Void
    ) -> Bool {
        var didAct = true
        if let identifier, PBGitDefaults.isDialogWarningSuppressed(forDialog: identifier) {
            actionBlock()
            return didAct
        }
        alert.showsSuppressionButton = true
        guard let window = windowController.window else { return didAct }
        alert.beginSheetModal(for: window) { response in
            guard response == .alertFirstButtonReturn else {
                didAct = false
                return
            }
            if let identifier, alert.suppressionButton?.state == .on {
                PBGitDefaults.suppressDialogWarning(forDialog: identifier)
            }
            actionBlock()
        }
        return didAct
    }
}
